Add some general guidelines around PR etiquette (#139)
canonical-webteam/practices: workflow/qa-steps.md

In this section you will find a list of steps to take when reviewing PRs, in order to catch unforeseen regressions.

## General guidelines for quality reviews

- Assume positive intent, we all have collective responsibility for the code we write - be nice.

> - Avoid expressing subjective opinions without reasoning. If you think something is bad, you need to explain _why_ you think it is bad.

Reviewer: Does _underscore_ do the same thing as *star*? I didn't know that :smile:

Author: It does - double underscore will also result in __bold__. No idea why they are two ways of doing it. 🤷🏻‍♂️

> - Try not to increase the scope of a PR unnecessarily. If a reviewing a PR leads you to unrelated code smells, create a new issue.

Reviewer: ```suggestion
- Try not to increase the scope of a PR unnecessarily. If reviewing a PR leads you to unrelated code smells, create a new issue.
```

Author: Yaaaas @nottrobin 🙌🏻

- Embrace the 'social coding' aspect of Github. If you spot a typo, rather than type a comment explaining where you found the typo and what the correction should be, [suggest the actual code change](https://help.github.com/articles/incorporating-feedback-in-your-pull-request/#applying-a-suggested-change).
- As above, you can also push refactors and larger code changes back to a PR which the original author can then review.

> - If you engage in a code review, you now have a responsibility to work with the author to get it over the line. Keep an eye out for your feedback being actioned and schedule a re-review in a timely manner.

Reviewer: I'm slightly worried about this point - I don't want anything to discourage people from jumping in and collaborating on PRs. If someone wants to jump into a PR to make one comment but that's all they've got time for, that seems valuable.

I'd rather we had powerful processed for pushing forward with PRs without relying on anyone in particular. For example, I tend to "dismiss" "changes requested" reviews once all the points are addressed. I think this is a good pattern. Another would be not to assume that because someone might have assigned themselves to a PR that you can't also look at it.

Or maybe we could formalise what "assigned" means, and say don't assign yourself unless you can commit. What do you think?

Author: I don't disagree with any of the above - my point here is that PRs shouldn't languish because it's waiting for a reviewer to return to it.

Reviewer: Great. So how about simply:

```suggestion
- If you engage in a code review, try to keep an eye out for your feedback being actioned so you can respond, to avoid blocking progress.
```

> - Have an awareness of team cadence. Software Engineering is an endless craft and we can all craft endlessly. Don't let 'perfect' become the enemy of 'good'. 

Reviewer: Is there an article or something that could illuminate this point a little better? As it is, I'm worried it may be a little too fuzzy. Does this mean I shouldn't point out issues where I see them? Where is the line?

Author: I agree it's a fuzzy point but I'm afraid it can't be anything but. 

The line has already always been there - we all decide at some stage when a review is complete.

Two articles which articulate this abstract point quite well;

[1] https://chris.charabaruk.com/2013/03/good-enough-versus-perfect/
[2] http://www.dudewho.codes/perfect-code/

Reviewer: Yes linking to one of those articles may help

I agree it's a good point for an individual developer to consider, but I think individuals would probably come to very different conclusions.

Some people might interpret this as meaning that it's better to be sloppy than spend the time to be more careful. Personally I think we should be striving to be pretty careful with everything we actually produce, but not procrastinate by radically limiting the scope (MVP basically).

At some point, I hope we can produce a set of overall guiding team principles here somewhere. A principle I've thought would be a good fit there would be "we care about the details". To some, that might very often conflict, in practice, with "perfect is the enemy of good".

I agree this is a good philosophy for people to be aware of, but if we're to make it a practice, I think we should make some effort to put a stake in the ground somewhere.

Could we maybe remove this one for now, pending further discussion? Or change it to something like:

```suggestion
- Avoid being overly pedantic. Always consider carefully if your point is important enough to hold up work being merged.
```

Author: As a perfect illustration of this ethos in action, I shall accept your compromise suggestion. 😁
